src/NHibernate/Linq/DefaultQueryProvider.cs

```diff
@@ -81,9 +88,19 @@ public TResult Execute<TResult>(Expression expression)
 			return (TResult)Execute(expression);
 		}
 
-		public void SetOptions(Action<NhQueryableOptions> setOptions)
+		public IQueryable<T> CreateQueryWithOptions<T>(IQueryable<T> queryable, Action<NhQueryableOptions> setOptions)
 		{
-			setOptions(_options.Value);
+			return WithOptions(setOptions)
+				.CreateQuery<T>(queryable.Expression);
+		}
+
+		public IQueryProvider WithOptions(Action<NhQueryableOptions> setOptions)
+		{
+			var options = _options != null
+				? _options.Clone()
+				: new NhQueryableOptions();
+			setOptions(options);
```

Comment:
We should guard against a `null` `setOptions` here, throwing `ArgumentNullException` instead of null ref.